oauth_dropins/reddit.py

```diff
@@ -0,0 +1,111 @@
+"""reddit OAuth drop-in.
+
+praw API docs:
+https://praw.readthedocs.io/en/v3.6.0/pages/oauth.html
+"""
+import logging
+
+from google.cloud import ndb
+import praw
+from webob import exc
+
+from . import handlers, models
+from .webutil import handlers as webutil_handlers
+from .webutil import util
+from .webutil.util import json_dumps, json_loads
+
+from random import randint
+
+REDDIT_APP_KEY = util.read('reddit_app_key')
+REDDIT_APP_SECRET = util.read('reddit_app_secret')
+
+class RedditAuth(models.BaseAuth):
+  """An authenticated reddit user.
+
+  Provides methods that return information about this user and make OAuth-signed
+  requests to the Tumblr API. Stores OAuth credentials in the datastore. See
+  models.BaseAuth for usage details.
+
+  reddit-specific details: implements "access_token," which is really a refresh_token
```

Comment:
oh ugh i see, their access tokens only last for 60m, so you have to refresh after that. ok, understood. https://github.com/reddit-archive/reddit/wiki/OAuth2#refreshing-the-token

except then the `access_token()` method should probably actually do the refresh and return a real access token, right? feel free to punt that to another PR if you want!

Reply:
conceptually returning the access_token makes sense, but after spending a while on it, I'm pretty sure praw doesn't give access to that directly.  instead they've written it so that you pass the refresh_token to the api every time.  do you know how i can change things to accommodate that? otherwise i guess i could write the authorize http request myself, but I'm not super familiar with how to do that.

Reply:
ah, got it! totally fine. exposing refresh token directly makes sense then, just name the method `refresh_token()`, and omit an `access_token()` method entirely. you can briefly mention which methods are available and not like this:

https://github.com/praw-dev/praw/blob/22c3d91c2a0d466bae5eeb555eb4e5892be53cab/praw/models/reddit/redditor.py#L41

Reply:
so i tried this originally, but it seems that access_token() is required by `oauth_dropins/models.py`.  I got a `NotImplementedError` when i remove that function.  Also it seems I can't have both a refresh_token attribute and a refresh_token method in the same class, should i just omit the method?

and did you mean to link to that table in praw? or were you trying to link to an example from this repo?

Reply:
hah sorry, yeah, misfired with the link.

ohhh is it this call that dies?

https://github.com/snarfed/oauth-dropins/blob/afe5cc946e2d969a37d83e2a093195c1d5cf27db/oauth_dropins/handlers.py#L203

it's not strictly necessary. maybe we wrapp it in a `try/catch NotImplementedError` and just omit the token in that case?

Reply:
OK, try catch on the NotImplementedError is in place.  Hopefully this wraps it up? Now onto the actual hard stuff :sweat_smile:

Reply:
looks like you maybe still need to `git add models.py`?

Reply:
i added the try except to the `handlers.py`. can you see that?

Reply:
oh! sorry! yes, got it now.